homeassistant/components/roborock/image.py

```diff
@@ -0,0 +1,130 @@
+
+from homeassistant.components.image import ImageEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import slugify
+import homeassistant.util.dt as dt_util
+
+from .const import DOMAIN
+from .coordinator import RoborockDataUpdateCoordinator
+from .device import RoborockCoordinatedEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Roborock image platform."""
+
+    coordinators: dict[str, RoborockDataUpdateCoordinator] = hass.data[DOMAIN][
+        config_entry.entry_id
+    ]
+    entities = list(
+        chain.from_iterable(
+            await asyncio.gather(
+                *(create_coordinator_maps(coord) for coord in coordinators.values())
+            )
+        )
+    )
+    async_add_entities(entities)
+
+
+class RoborockMap(RoborockCoordinatedEntity, ImageEntity):
+    """A class to let you visualize the map."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        unique_id: str,
+        coordinator: RoborockDataUpdateCoordinator,
+        map_flag: int,
+        starting_map: bytes,
+        map_name: str,
+    ) -> None:
+        """Initialize a Roborock map."""
+        RoborockCoordinatedEntity.__init__(self, unique_id, coordinator)
+        ImageEntity.__init__(self, coordinator.hass)
+        self._attr_name = map_name
+        drawables: list[Drawable] = [
+            Drawable.PATH,
+            Drawable.CHARGER,
+            Drawable.ROOM_NAMES,
+            Drawable.VACUUM_POSITION,
+        ]
+        self.parser = RoborockMapDataParser(
+            ColorsPalette(), Sizes(), drawables, ImageConfig(), []
+        )
+        self.async_update_token()
+        self._attr_image_last_updated = dt_util.utcnow()
+        self.map_flag = map_flag
+        self.cached_map = self._create_image(starting_map)
+
+    def should_update(self) -> bool:
+        """Update this map if it is the current active map, it's been long enough, and the vacuum is cleaning."""
+        return (
+            self.map_flag == self.coordinator.current_map
+            and self.image_last_updated is not None
+            and (dt_util.utcnow() - self.image_last_updated).total_seconds() > 90
+            and self.coordinator.roborock_device_info.props.status is not None
+            and self.coordinator.roborock_device_info.props.status.in_cleaning
+        )
+
+    async def async_image(self) -> bytes | None:
+        """Update the image if it is not cached."""
+        if self.should_update():
+            map_data: bytes = await self.cloud_api.get_map_v1()
+            self.cached_map = self._create_image(map_data)
+            self._attr_image_last_updated = dt_util.utcnow()
```

Comment:
Check out https://developers.home-assistant.io/docs/core/entity/image#image which says that the frontend will only fetch the image when this property changes, and it is not expected to update this from within this call.

Reply:
I don't think this has been addressed to follow the developer guidance.

Reply:
Ah I thought I handled this - my bad. But to be honest, I don't really get what the docs means is best practice in this case.

Reply:
I think it is saying that the data fetch should not be happening in the image serving path, but instead should happen when data and state is normally updated (e.g. https://developers.home-assistant.io/docs/integration_fetching_data, `async_update` etc).

So as a concrete example of what it is saying could be appropriate could be: from `_handle_coordinator_update` start another background task to fetch the cloud image, then as that finishes it sets the image update time.  That's just an example, not sure if its the best approach.

Reply:
Yeah is_cached_expired only updates the map when it is interesting.

I'm still a bit confused

`    @callback
    def _handle_coordinator_update(self) -> None:
        """Handle updated data from the coordinator."""
        if self.is_cache_expired():
            self._attr_image_last_updated = dt_util.utcnow()
        super()._handle_coordinator_update()

    async def async_image(self) -> bytes | None:
        """Update the image if it is not cached."""
        map_data: bytes = await self.cloud_api.get_map_v1()
        self.cached_map = self._create_image(map_data)
        self._attr_image_last_updated = dt_util.utcnow()
        return self.cached_map
`

But the issue is - the frontend will repeatedly call async_image if i refresh or open up the image, it doesn't seem to care about image_last_updated

Reply:
OK.. maybe its more about subscribing to updates and knowing when to refresh rather than to avoid refreshing too often. You may want to kick off the image fetch when the coordinator update finishes, then update the image and state at the same time.

(Overall the down side is the map will be getting refreshed when the robot is active even if nobody is looking, but that doesn't seem so bad.)

Reply:
I don't fully follow the docs logic though still

async_image gets called whenever I look at the image, if I handle updating the image there and also setting last_image_updated there, then it kind of solves all of the problems, no?

The map will get refreshed when someone looks at it and it needs to and the last_image_updated reflects when the image was last updated.

Perhaps I do what fritz is doing?

```
    async def async_update(self) -> None:
        """Update the image entity data."""
        qr_bytes = await self._fetch_image()

        if self._current_qr_bytes != qr_bytes:
            dt_now = dt_util.utcnow()
            _LOGGER.debug("qr code has changed, reset image last updated property")
            self._attr_image_last_updated = dt_now
            self._current_qr_bytes = qr_bytes
            self.async_write_ha_state()

    async def async_image(self) -> bytes | None:
        """Return bytes of image."""
        return self._current_qr_bytes

```

Reply:
It may be that the frontend is not yet taking advantage of this, but it is the currently specified API contract that it can if it needs to.

Yes, what fritz is doing is fine in general, but the complexity added in your case is the active map is determined by the update coordinator so I assume you'll want to run after it has the latest active map data.

Reply:
Should be resolved now